service-base/src/main/java/org/eclipse/hono/service/tenant/BaseExecutionContextTenantAndAuthIdProvider.java

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.service.tenant;
+
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.net.ssl.SSLSession;
+import javax.security.auth.x500.X500Principal;
+
+import org.eclipse.hono.client.TenantClientFactory;
+import org.eclipse.hono.config.ProtocolAdapterProperties;
+import org.eclipse.hono.util.Constants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+
+/**
+ * A base class for {@link ExecutionContextTenantAndAuthIdProvider} implementations.
+ */
+public class BaseExecutionContextTenantAndAuthIdProvider {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BaseExecutionContextTenantAndAuthIdProvider.class);
+
+    protected final ProtocolAdapterProperties config;
+    protected final TenantClientFactory tenantClientFactory;
+
+    /**
+     * Creates a new BaseExecutionContextTenantAndAuthIdProvider for the given config and tenantClientFactory.
+     *
+     * @param config The configuration.
+     * @param tenantClientFactory The factory to use for creating a Tenant service client.
+     */
+    public BaseExecutionContextTenantAndAuthIdProvider(final ProtocolAdapterProperties config,
+            final TenantClientFactory tenantClientFactory) {
+        this.config = config;
+        this.tenantClientFactory = tenantClientFactory;
+    }
+
+    /**
+     * Gets a {@link TenantObjectWithAuthId} from the X509 certificate of the given {@link SSLSession}.
+     *
+     * @param sslSession The SSL session.
+     * @param spanContext The OpenTracing context to use for tracking the operation.
+     * @return A future indicating the outcome of the operation.
+     */
+    protected Future<TenantObjectWithAuthId> getFromSslSessionCert(final SSLSession sslSession,
```

Comment:
this sounds quite technical. I'd say we are talking about the client cert, i.e. *getFromClientCertificate* ?